File: Codes/1695-maximum-erasure-value.cpp
/*
 * @Author: LetMeFly
 * @Date: 2025-07-23 10:31:42
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-07-23 13:26:19
 */
#if defined(_WIN32) || defined(__APPLE__)
#include "_[1,2]toVector.h"
#endif

class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        unordered_set<int> had;
        int ans = 0, cnt = 0;
        for (int l = 0, r = 0; r < nums.size(); r++) {
            while (had.count(nums[r])) {
                had.erase(nums[l]);
                cnt -= nums[l++];
            }
            had.insert(nums[r]);
            cnt += nums[r];
            // printf("l = %d, r = %d, cnt = %d\n", l, r, cnt);
            ans = max(ans, cnt);
        }
        return ans;
    }
};

/*
[187,470,25,436,538,809,441,167,477,110,275,133,666,345,411,459,490,266,987,965,429,166,809,340,467,318,125,165,809,610,31,585,970,306,42,189,169,743,78,810,70,382,367,490,787,670,476,278,775,673,299,19,893,817,971,458,409,886,434]
9252
16911
*/

File: Codes/1695-maximum-erasure-value.go
/*
 * @Author: LetMeFly
 * @Date: 2025-07-23 10:31:42
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-07-24 13:08:34
 */
package main

import "fmt"

// 昨天晚上干别的去了，今天中午补上，间隔时间更久效果更好(bushi)
func maximumUniqueSubarray(nums []int) (ans int) {
    had := map[int]struct{}{}
    l, cnt := 0, 0
    for _, t := range nums {
        for _, ok := had[t]; ok; {
            delete(had, nums[l])
            cnt -= nums[l]
            l++
        }
        cnt += t
        had[t] = struct{}{}
    }
    return
}

func main() {
    nums := []int{4, 2, 4, 5, 6}
    ans := maximumUniqueSubarray(nums)
    fmt.Println(ans)
}